rust/azure_iot_operations_protocol/src/rpc_command/invoker.rs

```diff
@@ -157,6 +159,357 @@ where
+                    aio_error.kind = AIOProtocolErrorKind::UnknownError;
+                    aio_error.property_name = value.invalid_property_name;
+                    // It is expected that value.invalid_property_value is None, but it is not set
+                    // in order to ensure a None value. This behavior comes from prior iterations
+                    // of code - unsure if this is truly desirable.
+                }
+            }
+            StatusCode::ServiceUnavailable => {
+                aio_error.kind = AIOProtocolErrorKind::StateInvalid;
+                aio_error.header_name = value.invalid_property_name;
+                aio_error.header_value = value.invalid_property_value;
+            }
+            StatusCode::VersionNotSupported => {
+                aio_error.kind = AIOProtocolErrorKind::UnsupportedVersion;
+            }
+        }
+        aio_error
+    }
+}
+
+/// Internal enum representing a result returned over the network
+enum CommandResult<TResp>
+where
+    TResp: PayloadSerialize,
+{
+    /// Indicates a successful response reported over the network
+    Ok(Response<TResp>),
+    /// Indicates an application-level failure reported over the network
+    Err(RemoteError),
+}
+
+impl<TResp> TryFrom<Publish> for CommandResult<TResp>
+where
+    TResp: PayloadSerialize,
+{
+    // NOTE: The `command_name` field will need to be added to the AIOProtocolError AFTER it is
+    // returned in the case of failure, since, of course, the command name is not included in the
+    // Publish.
+    // NOTE 2: This of course would make moving to other error models more complex, but the
+    // eventual implementation of a ChunkBuffer could get us around this by including the command
+    // name on the buffer.
+    type Error = AIOProtocolError;
+
+    fn try_from(value: Publish) -> Result<CommandResult<TResp>, Self::Error> {
+        // NOTE: User properties are parsed out into a new HashMap because:
+        // 1) It makes the code more readable/maintanable to do HashMap lookups
+        // 2) When this logic is extracted to a ChunkBuffer, it will be more memory efficient as
+        //  we won't want to keep entire copies of all Publishes, so we will just copy the
+        //  properties once.
+
+        let publish_properties =
+            value
+                .properties
+                .ok_or(AIOProtocolError::new_header_missing_error(
+                    "Properties",
+                    false,
+                    Some("Properties missing from MQTT message".to_string()),
+                    None,
+                ))?;
+
+        // Parse user properties
+        let expected_aio_properties = [
+            UserProperty::Timestamp,
+            UserProperty::Status,
+            UserProperty::StatusMessage,
+            UserProperty::IsApplicationError,
+            UserProperty::InvalidPropertyName,
+            UserProperty::InvalidPropertyValue,
+            UserProperty::ProtocolVersion,
```

Comment:
If the protocol version changes and that changes what fields we expect, will that be too disruptive to this flow? (for now, this flow definitely makes sense)

Reply:
This would have to updated, yes, but I think that's desirable because it's explicit, unlike before.